app/ts/contenthash.ts

```diff
@@ -0,0 +1,174 @@
+		}
+		if (!isCryptographicIPNS(cid)) throw new Error('not supported')
+		return CID.create(1, 0x72, cid.multihash).bytes
+	},
+	utf8: (value: string): Uint8Array => {
+		const encoder = new TextEncoder()
+		return encoder.encode(value)
+	},
+	arweave: (value: string): Uint8Array => base64Decode(value),
+}
+
+const decodes = {
+	hexMultiHash: (value: Uint8Array): string => {
+		const cid = CID.decode(value)
+		return bytesToHexString(multihashDecode(cid.multihash.bytes).digest)
+	},
+	ipfs: (value: Uint8Array): string => {
+		const cid = CID.decode(value).toV1()
+		return cid.toString(cid.code === 0x72 ? base36 : base32)
+	},
+	ipns: (value: Uint8Array): string => {
+		const cid = CID.decode(value).toV1()
+		if (!isCryptographicIPNS(cid)) throw new Error('not supported')
+		return cid.toString(base36)
+	},
+	utf8: (value: Uint8Array): string => {
+		const decoder = new TextDecoder()
+		return decoder.decode(value)
+	},
+	base64: (value: Uint8Array): string => {
+		return base64url.encode(value).substring(1)
+	},
+}
+
+export const encodeContentHash = (name: string, value: string): `0x${ string }` => {
+	const encode = () => {
+		switch(name) {
+			case 'arweave': return encodes.arweave(value)
+			case 'ipfs': return encodes.ipfs(value)
+			case 'ipns': return encodes.ipns(value)
+			case 'skynet': return encodes.skynet(value)
+			case 'swarm': return encodes.swarm(value)
+			default: return encodes.utf8(value)
+		}
+	}
+	const bytes = encode()
+	const code = nameToCode[name as keyof typeof nameToCode] as number
+	const codeBytes = varint.encodeTo(code, new Uint8Array(varint.encodingLength(code)))
+	return bytesToHexString(new Uint8Array([...codeBytes, ...bytes]))
+}
+
+export const decodeContentHash = (contentHash: string): string => {
+	const bytes = hexStringToBytes(contentHash)
+	const [code, offset] = varint.decode(bytes)
+	const value = bytes.slice(offset)
+	if (!(code in codeToName)) throw new Error('invalid protocol')
+	const name = codeToName[code as CodecId]
+	const withProtocol = (hash: string) => `${ name }://${ hash }`
+	switch(name) {
+		case 'arweave': return withProtocol(decodes.base64(value))
+		case 'ipfs': return withProtocol(decodes.ipfs(value))
+		case 'ipns': return withProtocol(decodes.ipns(value))
+		case 'onion': return withProtocol(decodes.utf8(value))
+		case 'onion3': return withProtocol(decodes.utf8(value))
+		case 'skynet': return withProtocol(decodes.base64(value))
+		case 'swarm': return withProtocol(decodes.hexMultiHash(value))
+		default: assertNever(name)
+	}
+}
+
+export const tryDecodeContentHash = (contentHash: string) => {
+	try {
+		return decodeContentHash(contentHash)
+	} catch(e) {
+		return undefined
+	}
+}
+
+export const tryEncodeContentHash = (contentHash: string) => {
+	try {
+		const [codec, hash] = contentHash.split('://')
+		if (codec === undefined || hash === undefined) throw new Error('invalid hash')
+		return encodeContentHash(codec, hash)
+	} catch(e) {
+		console.log(e)
```

Comment:
did you want to `console.warn` or `console.error` here?

Reply:
this is removed in later prs